### ISSUE TITLE
How to build on Windows? (No make command)

### ISSUE THREAD

**Issue by Alwinator** [ActivityWatch/activitywatch#489]
There is no "real" make command on windows. I have tried some tools which should add the make commands to Windows, but still can't build Activity Watch. Maybe you could make it CMake compartible.

Thanks a lot in advance.

**Comment by ErikBjare** [comment]
You can install GNU Make on Windows. That's what we do in CI. There are some basic instructions in the docs: https://docs.activitywatch.net/en/latest/installing-from-source-on-windows.html

**Comment by Alwinator** [comment]
Windows 10
latest master of activity watch
Python 3.7.3 installed and in the path variable
poetry installed and in the path variable
pyinstaller installed and in the path variable

```
PS C:\Users\alwin\Desktop\activitywatch-master> make build
if [ -e "aw-core/.git" ]; then \
                echo "Submodules seem to already be initialized, continuing..."; \
        else \
                git submodule update --init --recursive; \
        fi
-e was unexpected at this time.
make: *** [build] Error 255
```

**Comment by johan-bjareholt** [comment]
@Alwinator See this reply on the forum where someone else has the same issue

https://forum.activitywatch.net/t/parse-version-error/729/4?u=johan-bjareholt

**Comment by Alwinator** [comment]
I have added the path C:\Program Files\Git\usr\bin in system variables like in the forum described. It solved the first issue. It is also important to use git clone instead of downloading a zip file.

Then I got:
```
  - Installing aw-client (0.3.1)
make[1]: Leaving directory `C:/Users/alwin/Desktop/activitywatch/aw-client'
make --directory=aw-watcher-afk build
make[1]: Entering directory `C:/Users/alwin/Desktop/activitywatch/aw-watcher-afk'
poetry install
The lock file might not be compatible with the current version of Poetry.
Upgrade Poetry to ensure the lock file is read properly or, alternatively, regenerate the lock file with the `poetry lock` command.
Installing dependencies from lock file

[NonExistentKey]
'Key "reference" does not exist.'
make[1]: *** [build] Error 1
make[1]: Leaving directory `C:/Users/alwin/Desktop/activitywatch/aw-watcher-afk'
make: *** [build] Error 2
```

Fixed it by deleting the poetry.lock file in aw-watcher-afk.

It is also important to have [npm](https://www.npmjs.com/get-npm) and [cargo](https://doc.rust-lang.org/cargo/getting-started/installation.html) installed otherwise it will crash while building.

Then I got the following error:
```
error: failed to run custom build command for `rocket_codegen v0.4.5`

Caused by:
  process didn't exit successfully: `C:\Users\alwin\Desktop\activitywatch\aw-server-rust\target\release\build\rocket_codegen-e300c940ffc0d7af\build-script-build` (exit code: 101)
  --- stderr
  Error: Rocket (codegen) requires a 'dev' or 'nightly' version of rustc.
  Installed version: 1.46.0 (2020-08-24)
  Minimum required:  1.33.0-nightly (2019-01-13)
  See the getting started guide (https://rocket.rs/v0.5/guide/getting-started/) for more information.
  thread 'main' panicked at 'Aborting compilation due to incompatible compiler.', C:\Users\alwin\.cargo\registry\src\github.com-1ecc6299db9ec823\rocket_codegen-0.4.5\build.rs:29:13
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
make[1]: *** [aw-server] Error 101
make[1]: Leaving directory `C:/Users/alwin/Desktop/activitywatch/aw-server-rust'
make: *** [build] Error 2
```

I have fixed this error by using the following commands:
```
rustup install nightly
rustup default nightly
```

Then I got the following error:
```
  - Installing aw-core (0.4.2)
make[1]: Leaving directory `C:/Users/alwin/Desktop/activitywatch/aw-core'
python -c "import aw_server; print(aw_server.__version__)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'aw_server'
make: *** [build] Error 1
```
I fixed this error by running `make build` again in a virtual environment, because I forgot about this, because of the errors.....

Wait..., ohh there we go, next error:
```
  - Installing pyHook (1.5.1 https://github.com/ActivityWatch/wheels/raw/master/pyHook-1.5.1-cp37-cp37m-win_amd64.whl)

[EnvCommandError]
Command ['C:\\Users\\alwin\\Desktop\\activitywatch\\venv\\Scripts\\pip.exe', 'install', '--no-deps', 'https://github.com/ActivityWatch/wheels/raw/master/pyHook-1.5.1-cp37-cp37m-win_amd64.whl#egg=pyhook'] errored with the following return code 1, and output:
ERROR: pyHook-1.5.1-cp37-cp37m-win_amd64.whl is not a supported wheel on this platform.

make[1]: *** [build] Error 1
make[1]: Leaving directory `C:/Users/alwin/Desktop/activitywatch/aw-watcher-afk'
make: *** [build] Error 2
```
Fixed by changing line 18 from pyproject.toml in aw-watcher-afk from:
```
pyhook = {url = "https://github.com/ActivityWatch/wheels/raw/master/pyHook-1.5.1-cp37-cp37m-win_amd64.whl", platform = "win32" }
```
to
```
pyhook = {url = "https://github.com/ActivityWatch/wheels/raw/master/pyHook-1.5.1-cp36-cp36m-win_amd64.whl", platform = "win32" }
```
and line 12 from `python = "^3.7"` to `python = "^3.6"`

**make build succeeds now.**

```
pyinstaller .\aw.spec
```
Programmers must be patient (next error):
```
(venv) PS C:\Users\alwin\Desktop\activitywatch> pyinstaller .\aw.spec
61 INFO: PyInstaller: 3.5
61 INFO: Python: 3.7.3
62 INFO: Platform: Windows-10-10.0.18362-SP0
63 INFO: UPX is not available.
Traceback (most recent call last):
  File "c:\programdata\miniconda3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\programdata\miniconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\ProgramData\Miniconda3\Scripts\pyinstaller.exe\__main__.py", line 9, in <module>
  File "c:\programdata\miniconda3\lib\site-packages\PyInstaller\__main__.py", line 111, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "c:\programdata\miniconda3\lib\site-packages\PyInstaller\__main__.py", line 63, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "c:\programdata\miniconda3\lib\site-packages\PyInstaller\building\build_main.py", line 844, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "c:\programdata\miniconda3\lib\site-packages\PyInstaller\building\build_main.py", line 791, in build
    exec(code, spec_namespace)
  File ".\aw.spec", line 7, in <module>
    import aw_core
ModuleNotFoundError: No module named 'aw_core'
```
Fixed it by installing pyinstaller again:
```
pip install pyinstaller
```
(it was only installed globally)

Next error
```
Unable to find "aw-server-rust\target\package\aw-server-rust" when adding binary and data files
```
Fixed it by changing line 18 in aw.spec from:
```
aw_server_rust_bin = "aw-server-rust/target/package/aw-server-rust"
```
to
```
aw_server_rust_bin = "aw-server-rust/target/release"
```

So this is what I did after my school day.... I mean.. I hope it helps.

The build of Activity Watch is not very developer friendly e.g. that I have to build the rust server even if I don't use it or that I don't get an error if I don't have npm, cargo, ... installed.

**But everything works now and for such a great software like Activity Watch it is worth it.**

Now I have the dist folder from pyinstaller. How should I "install" it now? The content of the `C:\Program Files (x86)\ActivityWatch` looks pretty different. How did you build the Windows Installer?

**Comment by ErikBjare** [comment]
That is very helpful, thanks for being so thorough!

We should probably update the docs and build process with some of what you mentioned here to make it as easy as possible for people. If you find the time I'd gladly accept PRs :)

---

>Now I have the dist folder from pyinstaller. How should I "install" it now? The content of the C:\Program Files (x86)\ActivityWatch looks pretty different. How did you build the Windows Installer?

`make package` should build the installer since it calls `scripts/package/package-all.sh`, which in turn runs innosetup which builds the installer.

---

```
ERROR: pyHook-1.5.1-cp37-cp37m-win_amd64.whl is not a supported wheel on this platform.
```

What? But you're running Python 3.7? How could changing it to cp36 fix it? I'd expect that to break it.

I literally bumped our Windows CI to Python 3.7 and pyHook yesterday, so I'd really like to know if it's causing issues. Runs fine in CI though.

**Comment by johan-bjareholt** [comment]
Most of the issues you had is because you did not read the documentation, it mentions that you need to install npm, rust nightly, pyinstaller for "make package", virtualenv etc.

https://docs.activitywatch.net/en/latest/installing-from-source.html

There was a few things that were not specified in the docs though:
- It says that rust+cargo is optional while it in fact is not
- For some reason aw-server-rust did not run "make package" and only "make" (which is why it was placed in the wrong folder and you need to change it. I can't find the line you are refering to though at aw.spec:18?)
- pyhook for 3.7 not working, could you give us the log what the issue was?

**Comment by Alwinator** [comment]
> That is very helpful, thanks for being so thorough!
> 
> We should probably update the docs and build process with some of what you mentioned here to make it as easy as possible for people. If you find the time I'd gladly accept PRs :)

> Yeah, this should probably be checked early in the build process, or at least be better mentioned in the docs, sorry about that.

I'm glad that it helps :)
When I have time I will improve your Windows build documentation.


> `make package` should build the installer since it calls `scripts/package/package-all.sh`, which in turn runs innosetup which builds the installer.

Okay, thank you I will try it today afternoon.

> What? But you're running Python 3.7? How could changing it to cp36 fix it? I'd expect that to break it.
> 
> I literally bumped our Windows CI to Python 3.7 and pyHook yesterday, so I'd really like to know if it's causing issues. Runs fine in CI though.

> * pyhook for 3.7 not working, could you give us the log what the issue was?

Yes, I am running Python 3.7, but I just didn't work, but with downgrading to 3.6 it does. Not sure why. I think it was a version conflict on my side, because I have many Python versions installed. I think `poetry install` was executed with Python 3.6 instead of 3.7 for any reason.

> Naww, that means a lot! ❤️

I am glad! :)



> Most of the issues you had is mentioned in the documentation, it says that you need to install npm, rust nightly, pyinstaller for "make package", virtualenv etc.

Yeah, your are right I only read the "Installing from source (on Windows)" docs and didn't thought of it that there is also a general "Installing from source" documentation.



> * It says that rust+cargo is optional while it in fact is not
> * For some reason aw-server-rust did not run "make package" and only "make" (which is why it was placed in the wrong folder and you need to change it. I can't find the line you are refering to though at aw.spec:18?)
> * pyhook for 3.7 not working, could you give us the log what the issue was?

Thanks for summarizing the important parts.

> I can't find the line you are refering to though at aw.spec:18?)
https://github.com/ActivityWatch/activitywatch/blob/master/aw.spec#L18

**Comment by johan-bjareholt** [comment]
> Yeah, your are right I only read the "Installing from source (on Windows)" docs and didn't thought of it that there is also a general "Installing from source" documentation.

We could actually clear that up as well, start that page with saying that you also need to follow the general platform-agnostic guide.



>> I can't find the line you are refering to though at aw.spec:18?)
> https://github.com/ActivityWatch/activitywatch/blob/master/aw.spec#L18

Strange that I didn't find it after a git pull of the repo, will take a look at it later.

**Comment by johan-bjareholt** [comment]
Good catch about chocolatey being required for packaging.
It would probably be a better idea to instead only list innosetup as a dependency and change the script fail the build if it is not installed. That way the user has options to install it in other ways than through chocolatey.

**Comment by Alwinator** [comment]
Do you have an idea why the Activity Watch build does not work?

**Comment by johan-bjareholt** [comment]
I'm not sure how windows permissions work, but it sounds like a bad idea to run make package as an administrator. That's the only guess I have why it might make that permission error.

**Comment by ErikBjare** [comment]
@Alwinator Good catch about the manifest files, I'll fix it in aw-qt. 

Could you try a recent build from https://activitywatch-builds.s3.amazonaws.com/ (just append the filename for the most recent Windows build in the XML) and see if you get the same error?

It's possible that I fudged something up in a recent commit that I haven't yet tested on Windows.

**Comment by Alwinator** [comment]
> Could you try a recent build from https://activitywatch-builds.s3.amazonaws.com/ (just append the filename for the most recent Windows build in the XML) and see if you get the same error?

Sounds great, but the list does not include a Windows build.

**Comment by ErikBjare** [comment]
@Alwinator Ah sorry, forgot that only Travis-built stuff ends up there.

You can find builds [here](https://ci.appveyor.com/project/ErikBjare/activitywatch/build/artifacts), or some less-tested ones from [here](https://github.com/ActivityWatch/activitywatch/actions/runs/295547479).

If you wait a few hours from me making this comment I'll hopefully have merged https://github.com/ActivityWatch/aw-qt/pull/64 which should work better.

**Comment by Alwinator** [comment]
> @Alwinator Ah sorry, forgot that only Travis-built stuff ends up there.
> 
> You can find builds [here](https://ci.appveyor.com/project/ErikBjare/activitywatch/build/artifacts), or some less-tested ones from [here](https://github.com/ActivityWatch/activitywatch/actions/runs/295547479).
> 
> If you wait a few hours from me making this comment I'll hopefully have merged [ActivityWatch/aw-qt#64](https://github.com/ActivityWatch/aw-qt/pull/64) which should work better.

Thank you a lot! I have used the first link and it works. Also the "custom watcher loading system" works. And the thing with the Manifest files is fixed too.
